### DISCUSSION TITLE
NeighborLoader returns all empty except seed nodes

### DISCUSSION THREAD

**Discussion by cctony1** [pyg-team/pytorch_geometric#7270] [Q&A]
I'm able to use the neighbor loader with the ogb mag dataset in the example here: https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/to_hetero_mag.py

However, I have a problem for my own dataset.

I created this, train_data: 

HeteroData(
  p={
    x=[1517, 96],
    y=[1517]
  },
  mc={ x=[102354, 14] },
  rx={ x=[46659, 6] },
  po={ x=[2350, 36] },
  dx={ x=[3496, 22] },
  d={ x=[1993, 5] },
  pa={ x=[53022, 81] },
  (p, WP, rx)={ edge_index=[2, 46659] },
  (p, H, mc)={ edge_index=[2, 102354] },
  (rx, F, d)={ edge_index=[2, 46603] },
  (rx, PB, pa)={ edge_index=[2, 46659] },
  (mc, W, po)={ edge_index=[2, 60928] },
  (mc, W, dx)={ edge_index=[2, 41426] },
  (mc, SB, pa)={ edge_index=[2, 102354] }
)

But I'm having trouble getting this to work:
train_input_nodes = ('Person', torch.arange(3)) # just 3 nodes for checking code
kwargs = {'batch_size': 1, 'num_workers': 2, 'persistent_workers': True} # batch size of 1 for checking code
train_loader = NeighborLoader(train_data, num_neighbors=[2,2], shuffle=True, input_nodes=train_input_nodes, **kwargs)

The seed nodes are sampled correctly, but the other nodes are missing, as are edge indices. Here is the data from one batch:

HeteroData(
  p={
    x=[1, 96],
    y=[1],
    batch_size=1
  },
  mc={ x=[0, 14] },
  rx={ x=[0, 6] },
  po={ x=[0, 36] },
  dx={ x=[0, 22] },
  d={ x=[0, 5] },
  pa={ x=[0, 81] },
  (p, WP, rx)={ edge_index=[2, 0] },
  (p, H, mc)={ edge_index=[2, 0] },
  (rx, F, d)={ edge_index=[2, 0] },
  (rx, PB, pa)={ edge_index=[2, 0] },
  (mc, W, po)={ edge_index=[2, 0] },
  (mc, W, dx)={ edge_index=[2, 0] },
  (mc, SB, pa)={ edge_index=[2, 0] }
)

Is there any (maybe obvious?) reason why the other nodes and the edges aren't populated when the neighborhood is sampled? Is it necessarily my dataset, or is there an error in how I set up the loader?

Windows 10 Enterprise
torch                 1.12.1+cu113
torch-cluster         1.6.0
torch-geometric       2.0.4
torch-scatter         2.0.9
torch-sparse          0.6.15
torch-spline-conv     1.2.1
torchaudio            0.12.1+cu113
torchinfo             1.7.2
torchvision           0.13.1+cu113

**Comment by EdisonLeeeee** [comment]
> train_input_nodes = ('Person', torch.arange(3)) # just 3 nodes for checking code

Didn't see this type of node in your graph.

**Reply by cctony1** [reply]
Sorry. I had edited node and edge names for this post. In the real code, the first node label is Person. In other words, that edited line should read “… = (‘p’, torch.arange(3))”

**Reply by rusty1s** [reply]
The only way possible this can happen is if neither of your input nodes is connected to your graph/they are isolated. Might this be the reason?

**Reply by cctony1** [reply]
I don't think so.

If I look at the first 2 'p' nodes as seeds, they are connected to several 'mc' nodes:

source nodes...
train_data[('p', 'H', 'mc')].edge_index[0,:20]
  tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1])

destination nodes...
train_data[('p', 'H', 'mc')].edge_index[1,:20]
  tensor([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
         14, 380, 233, 311, 341, 330])

And the same can be shown for other edges from 'p'.

I really appreciate all your help!
I guess if I'm implementing the loader correctly, it has to be an issue with my dataset. I just haven't figured it out yet. 

I read all the nodes and edges from a neo4j graph and saved them into python objects. I then converted these all to tensors to match the structure of the (e.g., ogb_mag) hetero dataset (as I understand it). But perhaps I broke the links somehow when I created the edge_index tensors...

**Reply by cctony1** [reply]
But even if I somehow mixed up the edge_index tensors, wouldn't the source and destination nodes I pasted above result in connections (albeit erroneous) between nodes?

**Reply by cctony1** [reply]
Do the node or edge_index tensors have to be of a certain dtype?

**Reply by cctony1** [reply]
In my case:

p x: torch.float64
p y: torch.int32
mc: torch.uint8
rx: torch.uint8
po: torch.uint8
dx: torch.uint8
d: torch.uint8
pa: torch.int32

edges:
torch.int64
torch.int64
torch.int64
torch.int64
torch.int64
torch.int64
torch.int64

train_input_nodes[1]: torch.int64

**Reply by cctony1** [reply]
I just confirmed that the edge_index tensors match the raw data (pulled from neo4j), so I don't think that is the problem.
 
In desperation, I also changed the dtypes of the tensors to match those used in some examples (e.g., ogb_mag and also FakeDataset). But that made no difference.

**Reply by EdisonLeeeee** [reply]
Do you have a minimal example (if available) to reproduce? I'm happy to help with that.

**Reply by cctony1** [reply]
Let me put something together!

**Reply by cctony1** [reply]
```
import torch
from torch_geometric.data import HeteroData
from torch_geometric.loader import NeighborLoader

# data
minimal = HeteroData()
minimal['p'].x = torch.tensor([[0.0000, 0.0095], [1.0000, 0.0084]], dtype=torch.float32)
minimal['p'].y = torch.tensor([0, 1], dtype=torch.int64)
minimal['mc'].x = torch.tensor([[0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0],
                                [0, 0, 0, 1, 0]], dtype=torch.float32)
minimal[('p', 'H', 'mc')].edge_index = torch.tensor([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1,  1,  1,  1,  1],
        [ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 16, 17, 18, 19, 20]], dtype=torch.int64)
input_nodes = ('p', torch.arange(2))
kwargs = {'batch_size': 1}
num_neighbors = [3]

# loader
min_loader = NeighborLoader(minimal, num_neighbors=num_neighbors, shuffle=True,
                              input_nodes=input_nodes, 
                              **kwargs)

# check
for batch in min_loader:
    print('batch:', batch)
    print()
    print('batch p.x shape:', batch['p'].x.shape)
    print()
    print('batch p.x:', batch['p'].x)
    print()
    print('batch mc.x shape:', batch['mc'].x.shape)
    print()
    print('batch mc.x:', batch['mc'].x)
    print()
    print('batch edges shape:', batch[('p', 'H', 'mc')].edge_index.shape)
    print()
    print('batch edges:', batch[('p', 'H', 'mc')].edge_index)
    print('==========')
```

**Reply by EdisonLeeeee** [reply]
Thank you for your code. I can indeed reproduce your error. This seems a bug and I've opened an issue for it, see #7280

**Reply by cctony1** [reply]
Thanks. Any idea why it doesn’t happen for other datasets (that seem to have the same structure)? I can send an example of what I mean tomorrow…

**Reply by EdisonLeeeee** [reply]
I think the direction might have been reversed during sampling but I'm not quite sure. Could you try using `torch_geometric.transforms.ToUndirected` to see if it resolves your issue?

**Reply by cctony1** [reply]
It seems to work after making graph undirected!

**Reply by cctony1** [reply]
I now have NeighborLoader  working for this hetero data:

HeteroData(
  p={
    x=[7041, 96],
    y=[7041]
  },
  mc={ x=[466692, 15] },
  rx={ x=[215932, 6] },
  po={ x=[4034, 40] },
  dx={ x=[7264, 22] },
  d={ x=[3489, 5] },
  pa={ x=[230854, 81] },
  (rx, PB, pa)={ edge_index=[2, 215932] },
  (mc, SB, pa)={ edge_index=[2, 466692] },
  (mc, W, po)={ edge_index=[2, 271421] },
  (mc, W, dx)={ edge_index=[2, 195271] },
  (rx, F, d)={ edge_index=[2, 215713] },
  (p, H, mc)={ edge_index=[2, 466692] },
  (p, WP, rx)={ edge_index=[2, 215932] },
  (pa, rev_PB, rx)={ edge_index=[2, 215932] },
  (pa, rev_SB, mc)={ edge_index=[2, 466692] },
  (po, rev_W, mc)={ edge_index=[2, 271421] },
  (dx, rev_W, mc)={ edge_index=[2, 195271] },
  (d, rev_F, rx)={ edge_index=[2, 215713] },
  (mc, rev_H, p)={ edge_index=[2, 466692] },
  (rx, rev_WP, p)={ edge_index=[2, 215932] }
)

I'm new to PyG and GNNs, and let me know if I should ask this question elsewhere...

Now I am trying to use SAGEConv by transforming the model via to_hetero. However, I seem to remember reading that this transformation only works if all node features have the same dimensions (but I cannot find that in the docs anymore), otherwise it will delete the node features. Is that right? 

I ask because after using to_hetero, during the model.eval step of my first training epoch I'm getting this error:
RuntimeError: mat1 and mat2 shapes cannot be multiplied (436x36 and 40x256)

Here is the code:
```
lay1_channels = 256
lay2_channels = 256
lay3_channels = 256

model = Sequential('x, edge_index', [
    (SAGEConv((-1, -1), lay1_channels), 'x, edge_index -> x'),
    ReLU(inplace=True),
    (SAGEConv((-1, -1), lay2_channels), 'x, edge_index -> x'),
    ReLU(inplace=True),
    (SAGEConv((-1, -1), lay3_channels), 'x, edge_index -> x'),
    ReLU(inplace=True),
    (Linear(-1, 2), 'x -> x') # 2 output channels for binary classification
])
model = to_hetero(model, train_data.metadata(), aggr='sum').to(device)

@torch.no_grad()
def init_params():
    # Initialize lazy parameters via forwarding a single batch to the model:
    batch = next(iter(train_loader))
    batch = batch.to(device, 'edge_index')
    model(batch.x_dict, batch.edge_index_dict)

def train():
    model.train()

    total_examples = total_loss = 0
    for batch in tqdm(train_loader):
        optimizer.zero_grad()
        batch = batch.to(device, 'edge_index')
        batch_size = batch['p'].batch_size
        out = model(batch.x_dict, batch.edge_index_dict)['p'][:batch_size]
        loss = F.cross_entropy(out, batch['p'].y[:batch_size])
        loss.backward()
        optimizer.step()

        total_examples += batch_size
        total_loss += float(loss) * batch_size

    return total_loss / total_examples

@torch.no_grad()
def test(loader):
    model.eval()

    total_examples = total_correct = 0
    targets_forF1 = []
    outputs_forF1 = []
    for batch in tqdm(loader):
        batch = batch.to(device, 'edge_index')
        batch_size = batch['p'].batch_size
        out = model(batch.x_dict, batch.edge_index_dict)['p'][:batch_size]
        pred = out.argmax(dim=-1)

        total_examples += batch_size
        total_correct += int((pred == batch['p'].y[:batch_size]).sum())
        targets_forF1.append(batch['p'].y[:batch_size].cpu())
        outputs_forF1.append(pred.cpu())
        
    targets_forF1 = torch.cat(targets_forF1)
    outputs_forF1 = torch.cat(outputs_forF1)
    f1_batch = f1_score(outputs_forF1, targets_forF1, average='macro')
        
    return total_correct / total_examples, f1_batch

init_params()  # Initialize parameters
optimizer = torch.optim.Adam(model.parameters(), lr=0.001, weight_decay=0.0002) 

rslt_df = pd.DataFrame(columns=['Loss', 'Train_acc', 'Val_acc', 'Test_acc', 'Train_f1', 'Val_f1', 'Test_f1'])
best_val_acc = 0
start_patience = patience = 100
for epoch in range(1, 400):
    loss = train()
    train_acc, train_f1 = test(train_loader)
    val_acc, val_f1 = test(val_loader)
    test_acc, test_f1 = test(test_loader)
    print(f'Epoch: {epoch:04d}, Loss: {loss:.4f}, '
          f'Train: {train_acc:.4f}, Val: {val_acc:.4f}, Test: {test_acc:.4f}, '
          f'TrainF1: {train_f1:.4f}, ValF1: {val_f1:.4f}, TestF1: {test_f1:.4f})')
    
    rslt = pd.Series([loss, train_acc, val_acc, test_acc, train_f1, val_f1, test_f1],
                    index=['Loss', 'Train_acc', 'Val_acc', 'Test_acc', 'Train_f1', 'Val_f1', 'Test_f1'])
    rslt_df = pd.concat([rslt_df, rslt.to_frame().T], ignore_index=True)

    if best_val_acc <= val_acc:
        patience = start_patience
        best_val_acc = val_acc
    else:
        patience -= 1

    if patience <= 0:
        print('Stopping training as validation accuracy did not improve '
              f'for {start_patience} epochs')
        break
```

and error:


100%|██████████████████████████████████████████████████████████████████████████████████| 56/56 [00:06<00:00,  8.40it/s]
100%|██████████████████████████████████████████████████████████████████████████████████| 56/56 [00:04<00:00, 13.38it/s]
  0%|                                                                                           | 0/12 [00:02<?, ?it/s]
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
File <timed exec>:10, in <module>

File ~\Documents\.Envs\.venv_neo4j_01\lib\site-packages\torch\autograd\grad_mode.py:27, in _DecoratorContextManager.__call__.<locals>.decorate_context(*args, **kwargs)
     24 @functools.wraps(func)
     25 def decorate_context(*args, **kwargs):
     26     with self.clone():
---> 27         return func(*args, **kwargs)

Input In [8], in test(loader)
     53 batch = batch.to(device, 'edge_index')
     54 batch_size = batch['p'].batch_size
---> 55 out = model(batch.x_dict, batch.edge_index_dict)['p'][:batch_size]
     56 pred = out.argmax(dim=-1)
     58 total_examples += batch_size

File ~\Documents\.Envs\.venv_neo4j_01\lib\site-packages\torch\fx\graph_module.py:652, in GraphModule.recompile.<locals>.call_wrapped(self, *args, **kwargs)
    651 def call_wrapped(self, *args, **kwargs):
--> 652     return self._wrapped_call(self, *args, **kwargs)

File ~\Documents\.Envs\.venv_neo4j_01\lib\site-packages\torch\fx\graph_module.py:277, in _WrappedCall.__call__(self, obj, *args, **kwargs)
    275     raise e.with_traceback(None)
    276 else:
--> 277     raise e

File ~\Documents\.Envs\.venv_neo4j_01\lib\site-packages\torch\fx\graph_module.py:267, in _WrappedCall.__call__(self, obj, *args, **kwargs)
    265         return self.cls_call(obj, *args, **kwargs)
    266     else:
--> 267         return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
    268 except Exception as e:
    269     assert e.__traceback__

File ~\Documents\.Envs\.venv_neo4j_01\lib\site-packages\torch\nn\modules\module.py:1130, in Module._call_impl(self, *input, **kwargs)
   1126 # If we don't have any hooks, we want to skip the rest of the logic in
   1127 # this function, and just call forward.
   1128 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1129         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1130     return forward_call(*input, **kwargs)
   1131 # Do not call functions when jit is used
   1132 full_backward_hooks, non_full_backward_hooks = [], []

File <eval_with_key>.1:28, in forward(self, x, edge_index)
     26 module_0__pa1 = self.module_0.rx__PB__pa((x__rx, x__pa), edge_index__rx__PB__pa)
     27 module_0__pa2 = self.module_0.mc__SB__pa((x__mc, x__pa), edge_index__mc__SB__pa)
---> 28 module_0__po = self.module_0.mc__W__po((x__mc, x__po), edge_index__mc__W__po)
     29 module_0__dx = self.module_0.mc__W__dx((x__mc, x__dx), edge_index__mc__W__dx)
     30 module_0__d = self.module_0.rx__F__d((x__rx, x__d), edge_index__rx__F__d)

File ~\Documents\.Envs\.venv_neo4j_01\lib\site-packages\torch\nn\modules\module.py:1130, in Module._call_impl(self, *input, **kwargs)
   1126 # If we don't have any hooks, we want to skip the rest of the logic in
   1127 # this function, and just call forward.
   1128 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1129         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1130     return forward_call(*input, **kwargs)
   1131 # Do not call functions when jit is used
   1132 full_backward_hooks, non_full_backward_hooks = [], []

File ~\Documents\.Envs\.venv_neo4j_01\lib\site-packages\torch_geometric\nn\conv\sage_conv.py:85, in SAGEConv.forward(self, x, edge_index, size)
     83 x_r = x[1]
     84 if self.root_weight and x_r is not None:
---> 85     out += self.lin_r(x_r)
     87 if self.normalize:
     88     out = F.normalize(out, p=2., dim=-1)

File ~\Documents\.Envs\.venv_neo4j_01\lib\site-packages\torch\nn\modules\module.py:1130, in Module._call_impl(self, *input, **kwargs)
   1126 # If we don't have any hooks, we want to skip the rest of the logic in
   1127 # this function, and just call forward.
   1128 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1129         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1130     return forward_call(*input, **kwargs)
   1131 # Do not call functions when jit is used
   1132 full_backward_hooks, non_full_backward_hooks = [], []

File ~\Documents\.Envs\.venv_neo4j_01\lib\site-packages\torch_geometric\nn\dense\linear.py:118, in Linear.forward(self, x)
    113 def forward(self, x: Tensor) -> Tensor:
    114     r"""
    115     Args:
    116         x (Tensor): The features.
    117     """
--> 118     return F.linear(x, self.weight, self.bias)

RuntimeError: mat1 and mat2 shapes cannot be multiplied (436x36 and 40x256)

**Reply by cctony1** [reply]
((perhaps I was thinking of HeteroData().to_homogeneous when it comes to feature dimensionality))

Regardless, I appreciate your insight about the mat1 x mat2 error.

**Reply by cctony1** [reply]
It looks like the error goes away if I remove the test(val_loader) and test(test_loader) lines.

I realize my val/test sets have a few nodes with slightly different feature dimensions from train (probably an oversight when I onehot encoded some variables in val/test that were missing categories that existed in train...)

**Reply by cctony1** [reply]
I fixed the variable issue. Training+Eval is working now. But I get validation and test accuracies of 100% (haha, wow)

I guess I'll look into data leakage. Otherwise, do you see any issue with my GNN?

**Reply by EdisonLeeeee** [reply]
No. Your GNN model looks 100% correct to me.

**Reply by cctony1** [reply]
Thanks! It's working now like a normal model (found error that led to perfect accuracy...)

**Reply by cctony1** [reply]
Since we're going from Destination to Source...

when I specify num_neighbors=[10,20] for the NeighborLoader, does the 10 or 20 refer to the seed nodes' nearest neighbors?

I had assumed 10 was nearest neighbors to seed, and 20 was the neighbors' neighbors. But since I had the message passing direction reversed in my head, I want to verify this detail.
